optuna/batch/study.py

```diff
@@ -0,0 +1,197 @@
+from typing import Union
+
+import numpy as np
+
+import optuna
+from optuna._experimental import experimental
+import optuna.logging
+
+ObjectiveFuncType = Callable[["optuna.batch.trial.BatchTrial"], np.ndarray]
+CallbackFuncType = Callable[["optuna.study.Study", "optuna.trial.FrozenTrial"], None]
+
+_logger = optuna.logging.get_logger(__name__)
+
+
+class BatchStudy(object):
+    def __init__(self, study: "optuna.study.Study", batch_size: int):
+        self._study = study
+        self._batch_size = batch_size
+
+    def __getattr__(self, attr_name: str) -> Any:
+        return getattr(self._study, attr_name)
+
+    @property
+    def batch_size(self) -> int:
+        """Return the size of batches.
+
+        Returns:
+            Size of batches.
+        """
+
+        return self._batch_size
+
+    def optimize(
+        self,
+        objective: ObjectiveFuncType,
+        timeout: Optional[int] = None,
+        n_batches: Optional[int] = None,
+        n_jobs: int = 1,
+        catch: Tuple[Type[Exception], ...] = (),
+        callbacks: Optional[List[CallbackFuncType]] = None,
+        gc_after_trial: bool = True,
+        show_progress_bar: bool = False,
+    ) -> None:
+
+        n_trials = math.ceil(n_batches / n_jobs) if n_batches is not None else None
+
+        self._study._run_trial_and_callbacks = types.MethodType(  # type: ignore
+            partial(_run_trial_and_callbacks, batch_func=objective, batch_size=self._batch_size),
+            self._study,
+        )
+        self._study.optimize(
+            lambda _: 0,
+            timeout=timeout,
+            n_trials=n_trials,
+            catch=catch,
+            callbacks=callbacks,
+            gc_after_trial=gc_after_trial,
+            show_progress_bar=show_progress_bar,
+        )
+
+
+def _run_trial_and_callbacks(
+    self: "optuna.study.Study",
+    func: "optuna.study.ObjectiveFuncType",
+    catch: Tuple[Type[Exception], ...],
+    callbacks: Optional[List[Callable[["optuna.study.Study", "optuna.trial.FrozenTrial"], None]]],
+    gc_after_trial: bool,
+    batch_func: ObjectiveFuncType,
+    batch_size: int,
+) -> None:
+
+    trials = _run_trial(self, batch_func, catch, gc_after_trial, batch_size)
+    if callbacks is None:
+        return
+
+    for trial in trials:
+        frozen_trial = copy.deepcopy(self._storage.get_trial(trial._trial_id))
+        for callback in callbacks:
+            callback(self, frozen_trial)
+
+
+def _run_trial(
+    self: "optuna.study.Study",
+    func: ObjectiveFuncType,
+    catch: Tuple[Type[Exception], ...],
+    gc_after_trial: bool,
+    batch_size: int,
+) -> List["optuna.trial.Trial"]:
+
+    # Sync storage once at the beginning of the objective evaluation.
+    self._storage.read_trials_from_remote_storage(self._study_id)
+
+    trials = []
+    for _ in range(batch_size):
+        trial_id = self._pop_waiting_trial_id()
+        if trial_id is None:
+            trial_id = self._storage.create_new_trial(self._study_id)
```

Comment:
Not really a race condition but trial numbers within a single batch will not necessarily follow a sequence under distributed optimization. This might be suboptimal when analyzing studies with e.g. visualization when you might want to clearly see the boundaries between batches. I'm wondering if we cannot creates trials in batches as well.

Reply:
> This might be suboptimal when analyzing studies with e.g. visualization when you might want to clearly see the boundaries between batches.

Can we provide such features as a post-process?
For example, 
1. Set the batch id to the 'system_attr' of each `Trial` instance,
2. After the optimization, a function sort the trials based on the batch id in the system attr and create a new study for analysis

Assigning the consecutive numbers to the trials in a batch is ideal, but I feel it might be overengineering because it requires new storage API.